File: mg_custom_nodes/ZhiHui6_zhihui_nodes_comfyui_20260111/Nodes/ImageAspectRatio.py
class ImageAspectRatio:
    PRESETS_QWEN = {
         "1:1(1328x1328)": (1328, 1328),
         "16:9(1664x928)": (1664, 928),
         "9:16(928x1664)": (928, 1664),
         "4:3(1472x1104)": (1472, 1104),
         "3:4(1104x1472)": (1104, 1472),
         "3:2(1584x1056)": (1584, 1056),
         "2:3(1056x1584)": (1056, 1584),
    }

    PRESETS_FLUX = {
         "1:1(1024x1024)": (1024, 1024),
         "16:9(1344x768)": (1344, 768),
         "9:16(768x1344)": (768, 1344),
         "4:3(1152x864)": (1152, 864),
         "3:4(864x1152)": (864, 1152),
         "3:2(1216x832)": (1216, 832),
         "2:3(832x1216)": (832, 1216),
    }
    
    PRESETS_WAN = {
         "1:1(1024x1024)": (1024, 1024),
         "16:9(1280x720)": (1280, 720),
         "9:16(720x1280)": (720, 1280),
         "4:3(1024x768)": (1024, 768),
         "3:4(768x1024)": (768, 1024),
         "21:9(1344x576)": (1344, 576),
         "9:21(576x1344)": (576, 1344),
    }
    
    PRESETS_SD = {
         "1:1(1024x1024)": (1024, 1024),
         "9:8(1152x896)": (1152, 896),
         "8:9(896x1152)": (896, 1152),
         "3:2(1216x832)": (1216, 832),
         "2:3(832x1216)": (832, 1216),
         "7:4(1344x768)": (1344, 768),
         "4:7(768x1344)": (768, 1344),
         "12:5(1536x640)": (1536, 640),
         "5:12(640x1536)": (640, 1536),
    }

    PRESETS_ZIMAGE = {
         "1:1(1024x1024)": (1024, 1024),
         "1:1(1280x1280)": (1280, 1280),
         "1:1(1536x1536)": (1536, 1536),
         "9:7(1152x896)": (1152, 896),
         "9:7(1440x1120)": (1440, 1120),
         "9:7(1728x1344)": (1728, 1344),
         "7:9(896x1152)": (896, 1152),
         "7:9(1120x1440)": (1120, 1440),
         "7:9(1344x1728)": (1344, 1728),
         "4:3(1152x864)": (1152, 864),
         "4:3(1472x1104)": (1472, 1104),
         "4:3(1728x1296)": (1728, 1296),
         "3:4(864x1152)": (864, 1152),
         "3:4(1104x1472)": (1104, 1472),
         "3:4(1296x1728)": (1296, 1728),
         "3:2(1248x832)": (1248, 832),
         "3:2(1536x1024)": (1536, 1024),
         "3:2(1872x1248)": (1872, 1248),
         "2:3(832x1248)": (832, 1248),
         "2:3(1024x1536)": (1024, 1536),
         "2:3(1248x1872)": (1248, 1872),
         "16:9(1280x720)": (1280, 720),
         "16:9(1536x864)": (1536, 864),
         "16:9(2048x1152)": (2048, 1152),
         "9:16(720x1280)": (720, 1280),
         "9:16(864x1536)": (864, 1536),
         "9:16(1152x2048)": (1152, 2048),
         "21:9(1344x576)": (1344, 576),
         "21:9(1680x720)": (1680, 720),
         "21:9(2016x864)": (2016, 864),
         "9:21(576x1344)": (576, 1344),
         "9:21(720x1680)": (720, 1680),
         "9:21(864x2016)": (864, 2016),
    }

    @classmethod
    def INPUT_TYPES(s):
        all_aspect_ratios = list(set(
            list(s.PRESETS_QWEN.keys()) +
            list(s.PRESETS_FLUX.keys()) +
            list(s.PRESETS_WAN.keys()) +
            list(s.PRESETS_SD.keys()) +
            list(s.PRESETS_ZIMAGE.keys())
        ))

        return {
            "required": {
                "preset_mode": (["Qwen image", "Flux", "Wan", "SDXL", "Z-image", "Custom Size"], {"default": "Z-image"}),
                "aspect_ratio": (all_aspect_ratios, {"default": "1:1(1328x1328)"}),
            },
            "optional": {
                "custom_width": ("INT", {"default": 1328, "min": 1, "max": 8192, "step": 1}),
                "custom_height": ("INT", {"default": 1328, "min": 1, "max": 8192, "step": 1}),
                "aspect_lock": ("BOOLEAN", {"default": False}),
            }
        }

    RETURN_TYPES = ("INT", "INT")
    RETURN_NAMES = ("width", "height")
    FUNCTION = "get_dimensions"
    CATEGORY = "Zhi.AI/Image"
    DESCRIPTION = "Smart image aspect ratio setting node, supporting preset dimensions for multiple models like Qwen, Flux, Wan, SDXL, with quick selection of common ratios or custom sizes"

    def get_dimensions(self, preset_mode, aspect_ratio, custom_width=1328, custom_height=1328, aspect_lock=False):
        if custom_width is None:
            custom_width = 1328
        if custom_height is None:
            custom_height = 1328

        if preset_mode == "Custom Size":
            return (custom_width, custom_height)

        preset_maps = {
            "Qwen image": self.PRESETS_QWEN,
            "Flux": self.PRESETS_FLUX,
            "Wan": self.PRESETS_WAN,
            "SDXL": self.PRESETS_SD,
            "Z-image": self.PRESETS_ZIMAGE
        }

        preset_map = preset_maps.get(preset_mode, self.PRESETS_QWEN)

        if aspect_ratio in preset_map:
            return preset_map[aspect_ratio]
        else:
            return (custom_width, custom_height)

File: mg_custom_nodes/ZhiHui6_zhihui_nodes_comfyui_20260129/web/ImageAspectRatio.js
import { app } from "../../../scripts/app.js";

app.registerExtension({
    name: "ImageAspectRatio",
    async beforeRegisterNodeDef(nodeType, nodeData, app) {
        if (nodeData.name === "ImageAspectRatio") {
            const onNodeCreated = nodeType.prototype.onNodeCreated;
            nodeType.prototype.onNodeCreated = function () {
                const r = onNodeCreated ? onNodeCreated.apply(this, arguments) : undefined;
                
                this.initialized = true;
                
                this.lastCustomSize = { width: 1328, height: 1328 };
                this.lastPresetSize = null;

                const QWEN_PRESETS = [
                    "1:1(1328x1328)",
                    "16:9(1664x928)",
                    "9:16(928x1664)",
                    "4:3(1472x1104)",
                    "3:4(1104x1472)",
                    "3:2(1584x1056)",
                    "2:3(1056x1584)",
                ];
                const FLUX_PRESETS = [
                    "1:1(1024x1024)",
                    "16:9(1344x768)",
                    "9:16(768x1344)",
                    "4:3(1152x864)",
                    "3:4(864x1152)",
                    "3:2(1216x832)",
                    "2:3(832x1216)",
                ];

                const FLUX2_PRESETS = [
                    "1:1·S(1024x1024)",
                    "1:1·M(1280x1280)",
                    "1:1·L(1536x1536)",

                    "16:9·S(1344x768)",
                    "16:9·M(1728x960)",
                    "16:9·L(2048x1152)",

                    "9:16·S(768x1344)",
                    "9:16·M(960x1728)",
                    "9:16·L(1152x2048)",

                    "4:3·S(1152x864)",
                    "4:3·M(1408x1056)",
                    "4:3·L(1728x1296)",

                    "3:4·S(864x1152)",
                    "3:4·M(1088x1472)",
                    "3:4·L(1280x1728)",

                    "3:2·S(1216x832)",
                    "3:2·M(1536x1024)",
                    "3:2·L(1920x1280)",

                    "2:3·S(832x1216)",
                    "2:3·M(1024x1536)",
                    "2:3·L(1280x1920)",
                ];

                const FLUX2_KLEIN_PRESETS = [
                    "1:1·S(768x768)",
                    "1:1·M(896x896)",
                    "1:1·L(1024x1024)",

                    "16:9·S(1024x576)",
                    "16:9·M(1152x640)",
                    "16:9·L(1344x768)",

                    "9:16·S(576x1024)",
                    "9:16·M(640x1152)",
                    "9:16·L(768x1344)",

                    "4:3·S(768x576)",
                    "4:3·M(960x704)",
                    "4:3·L(1024x768)",

                    "3:4·S(576x768)",
                    "3:4·M(704x960)",
                    "3:4·L(768x1024)",

                    "3:2·S(960x640)",
                    "3:2·M(1152x768)",
                    "3:2·L(1344x896)",

                    "2:3·S(640x960)",
                    "2:3·M(768x1152)",
                    "2:3·L(896x1344)",
                ];

                const WAN_PRESETS = [
                    "1:1(1024x1024)",
                    "16:9(1280x720)",
                    "9:16(720x1280)",
                    "4:3(1024x768)",
                    "3:4(768x1024)",
                    "21:9(1344x576)",
                    "9:21(576x1344)",
                ];
                const SD_PRESETS = [
                    "1:1(1024x1024)",
                    "9:8(1152x896)",
                    "8:9(896x1152)",
                    "3:2(1216x832)",
                    "2:3(832x1216)",
                    "7:4(1344x768)",
                    "4:7(768x1344)",
                    "12:5(1536x640)",
                    "5:12(640x1536)",
                    "9:16(768x1344)",
                    "16:9(1344x768)",
                ];
                const ZIMAGE_PRESETS = [
                    "1:1(1024x1024)",
                    "1:1(1280x1280)",
                    "1:1(1536x1536)",
                    "9:7(1152x896)",
                    "9:7(1440x1120)",
                    "9:7(1728x1344)",
                    "7:9(896x1152)",
                    "7:9(1120x1440)",
                    "7:9(1344x1728)",
                    "4:3(1152x864)",
                    "4:3(1472x1104)",
                    "4:3(1728x1296)",
                    "3:4(864x1152)",
                    "3:4(1104x1472)",
                    "3:4(1296x1728)",
                    "3:2(1248x832)",
                    "3:2(1536x1024)",
                    "3:2(1872x1248)",
                    "2:3(832x1248)",
                    "2:3(1024x1536)",
                    "2:3(1248x1872)",
                    "16:9(1280x720)",
                    "16:9(1536x864)",
                    "16:9(2048x1152)",
                    "9:16(720x1280)",
                    "9:16(864x1536)",
                    "9:16(1152x2048)",
                    "21:9(1344x576)",
                    "21:9(1680x720)",
                    "21:9(2016x864)",
                    "9:21(576x1344)",
                    "9:21(720x1680)",
                    "9:21(864x2016)",
                ];


                const setComboOptions = (widget, values) => {
                    if (!widget) return;
                    if (Array.isArray(widget.options)) {
                        widget.options = values;
                    } else if (widget.options && Array.isArray(widget.options.values)) {
                        widget.options.values = values;
                    } else {
                        widget.options = values;
                    }
                };


                const applyAspectOptionsByMode = (mode, aspectRatioWidget) => {
                    let presets;
                    switch (mode) {
                        case "Flux":
                            presets = FLUX_PRESETS;
                            break;
                        case "Flux.2":
                            presets = FLUX2_PRESETS;
                            break;
                        case "Flux2 klein":
                            presets = FLUX2_KLEIN_PRESETS;
                            break;
                        case "Wan":
                            presets = WAN_PRESETS;
                            break;
                        case "SDXL":
                            presets = SD_PRESETS;
                            break;
                        case "Z-image":
                            presets = ZIMAGE_PRESETS;
                            break;
                        case "Custom Size":
                            presets = [];
                            break;
                        case "Qwen image":
                        default:
                            presets = QWEN_PRESETS;
                            break;
                    }
                    const opts = [...presets];
                    setComboOptions(aspectRatioWidget, opts);
                    if (opts.length > 0 && !opts.includes(aspectRatioWidget.value)) {
                        aspectRatioWidget.value = opts[0];
                    }
                };

                this.swapWidthHeight = function() {
                    const widthWidget = this.widgets.find(w => w.name === "custom_width");
                    const heightWidget = this.widgets.find(w => w.name === "custom_height");
                    
                    if (widthWidget && heightWidget) {
                        const currentWidth = parseInt(widthWidget.value) || 1328;
                        const currentHeight = parseInt(heightWidget.value) || 1328;
                        
                        widthWidget.value = currentHeight;
                        heightWidget.value = currentWidth;
                        
                        this.lastCustomSize = {
                            width: currentHeight,
                            height: currentWidth
                        };
                        
                        if (this.aspectRatio && this.aspectRatio > 0) {
                            this.aspectRatio = currentHeight / currentWidth;
                            this.lastWidth = currentHeight;
                            this.lastHeight = currentWidth;
                        }
                        
                        app.graph.setDirtyCanvas(true, true);
                    }
                };

                this.repositionCustomControls = function(isCustom) {
                    const currentSize = { ...this.size };
                    
                    const widgets = this.widgets || [];
                    const modeIdx = widgets.findIndex(w => w && w.name === "preset_mode");
                    const aspectIdx = widgets.findIndex(w => w && w.name === "aspect_ratio");
                    const widthIdx = widgets.findIndex(w => w && w.name === "custom_width");
                    const heightIdx = widgets.findIndex(w => w && w.name === "custom_height");
                    const lockIdx = widgets.findIndex(w => w && w.name === "aspect_lock");
                    const swapIdx = widgets.findIndex(w => w && w.name === "🔁互换宽高·Swap_W&H");
                    
                    if (modeIdx < 0 || aspectIdx < 0 || widthIdx < 0 || heightIdx < 0 || lockIdx < 0) return;
                    
                    const aspectWidget = widgets[aspectIdx];
                    const widthWidget = widgets[widthIdx];
                    const heightWidget = widgets[heightIdx];
                    const lockWidget = widgets[lockIdx];
                    let swapWidget = swapIdx >= 0 ? widgets[swapIdx] : null;
                    
                    if (!isCustom) {
                        this.lastCustomSize = {
                            width: parseInt(widthWidget.value) || 1328,
                            height: parseInt(heightWidget.value) || 1328
                        };
                    }
                    
                    const indicesToRemove = [aspectIdx, widthIdx, heightIdx, lockIdx, swapIdx].filter(idx => idx >= 0).sort((a, b) => b - a);
                    for (const idx of indicesToRemove) {
                        widgets.splice(idx, 1);
                    }
                    
                    if (isCustom) {
                        widgets.splice(modeIdx + 1, 0, widthWidget, heightWidget);
                        
                        if (!swapWidget) {
                            swapWidget = this.addWidget("button", "🔁互换宽高·Swap_W&H", null, () => {
                                this.swapWidthHeight();
                            });
                            const curIdx = widgets.indexOf(swapWidget);
                            if (curIdx >= 0) widgets.splice(curIdx, 1);
                        }

                        widgets.splice(modeIdx + 3, 0, swapWidget, lockWidget);
                        
                        aspectWidget.hidden = true;
                        widthWidget.hidden = false;
                        heightWidget.hidden = false;
                        lockWidget.hidden = false;
                        if (swapWidget) {
                            swapWidget.hidden = false;
                        }
                        widthWidget.value = this.lastCustomSize.width;
                        heightWidget.value = this.lastCustomSize.height;
                        widgets.push(aspectWidget);
                    } else {
                        widgets.splice(modeIdx + 1, 0, aspectWidget);
                        aspectWidget.hidden = false;
                        widthWidget.hidden = true;
                        heightWidget.hidden = true;
                        lockWidget.hidden = true;
                        widgets.push(widthWidget, heightWidget, lockWidget);
                        if (swapWidget) {
                            swapWidget.hidden = true;
                            widgets.push(swapWidget);
                        }
                    }
                    
                    this.size = currentSize;
                    app.graph.setDirtyCanvas(true, true);
                };

                const originalCallback = this.callback;
                this.callback = function() {
                    if (originalCallback) {
                        originalCallback.apply(this, arguments);
                    }
                    
                    const aspectRatioWidget = this.widgets.find(w => w.name === "aspect_ratio");
                    const widthWidget = this.widgets.find(w => w.name === "custom_width");
                    const heightWidget = this.widgets.find(w => w.name === "custom_height");
                    const lockWidget = this.widgets.find(w => w.name === "aspect_lock");
                    const modeWidget = this.widgets.find(w => w.name === "preset_mode");
                    
                    if (!aspectRatioWidget || !widthWidget || !heightWidget || !lockWidget || !modeWidget) {
                        return;
                    }
                    
                    applyAspectOptionsByMode(modeWidget.value, aspectRatioWidget);

                    const isCustom = modeWidget.value === "Custom Size";
                    
                    this.repositionCustomControls(isCustom);
                    
                    if (!isCustom) {
                        lockWidget.value = false;
                        this.aspectRatio = null;
                        return;
                    }
                    
                    if (widthWidget.value === '' || widthWidget.value === null || widthWidget.value === undefined) {
                        widthWidget.value = 1328;
                    }
                    if (heightWidget.value === '' || heightWidget.value === null || heightWidget.value === undefined) {
                        heightWidget.value = 1328;
                    }
                    
                    this.lastWidth = parseInt(widthWidget.value) || 1328;
                    this.lastHeight = parseInt(heightWidget.value) || 1328;
                };
                
                setTimeout(() => {
                    const widthWidget = this.widgets.find(w => w.name === "custom_width");
                    const heightWidget = this.widgets.find(w => w.name === "custom_height");
                    const aspectRatioWidget = this.widgets.find(w => w.name === "aspect_ratio");
                    const lockWidget = this.widgets.find(w => w.name === "aspect_lock");
                    const modeWidget = this.widgets.find(w => w.name === "preset_mode");
                    
                    if (widthWidget && heightWidget && aspectRatioWidget && lockWidget && modeWidget) {
                        if (widthWidget.value === '' || widthWidget.value === null || widthWidget.value === undefined) {
                            widthWidget.value = 1328;
                        }
                        if (heightWidget.value === '' || heightWidget.value === null || heightWidget.value === undefined) {
                            heightWidget.value = 1328;
                        }
                        
                        this.lastWidth = parseInt(widthWidget.value) || 1328;
                        this.lastHeight = parseInt(heightWidget.value) || 1328;
                        this.aspectRatio = this.lastHeight > 0 ? this.lastWidth / this.lastHeight : 1;
                        this.isUpdating = false;
                        
                        const originalAspectCallback = aspectRatioWidget.callback;
                        const originalLockCallback = lockWidget.callback;
                        const originalModeCallback = modeWidget.callback;
                        
                        aspectRatioWidget.callback = () => {
                            if (originalAspectCallback) originalAspectCallback.apply(aspectRatioWidget, arguments);
                            this.callback();
                        };
                        
                        lockWidget.callback = () => {
                            if (originalLockCallback) originalLockCallback.apply(lockWidget, arguments);
                            if (lockWidget.value) {
                                const currentWidth = parseInt(widthWidget.value) || 1328;
                                const currentHeight = parseInt(heightWidget.value) || 1328;
                                if (currentHeight > 0) {
                                    this.aspectRatio = currentWidth / currentHeight;
                                    this.lastWidth = currentWidth;
                                    this.lastHeight = currentHeight;
                                } else {
                                    this.aspectRatio = 1;
                                    this.lastWidth = 1328;
                                    this.lastHeight = 1328;
                                }
                            }
                            this.lastCustomSize = {
                                width: parseInt(widthWidget.value) || 1328,
                                height: parseInt(heightWidget.value) || 1328
                            };
                            this.callback();
                        };

                        modeWidget.callback = () => {
                            if (originalModeCallback) originalModeCallback.apply(modeWidget, arguments);
                            applyAspectOptionsByMode(modeWidget.value, aspectRatioWidget);
                            
                            const isCustom = modeWidget.value === "Custom Size";
                            this.repositionCustomControls(isCustom);
                            
                            this.callback();
                        };
                        
                        this.aspectRatioTimer = setInterval(() => {
                            if (lockWidget.value && modeWidget.value === "Custom Size") {
                                const currentWidth = parseInt(widthWidget.value) || 0;
                                const currentHeight = parseInt(heightWidget.value) || 0;
                                
                                if (!this.aspectRatio && currentWidth > 0 && currentHeight > 0) {
                                    this.aspectRatio = currentWidth / currentHeight;
                                    this.lastWidth = currentWidth;
                                    this.lastHeight = currentHeight;
                                }
                                
                                if (this.aspectRatio && (currentWidth !== this.lastWidth || currentHeight !== this.lastHeight)) {
                                    if (!this.isUpdating) {
                                        this.isUpdating = true;
                                        
                                        if (currentWidth !== this.lastWidth && currentWidth > 0) {
                                            const newHeight = Math.round(currentWidth / this.aspectRatio);
                                            if (!isNaN(newHeight) && newHeight > 0) {
                                                heightWidget.value = newHeight;
                                                this.lastHeight = newHeight;
                                            }
                                            this.lastWidth = currentWidth;
                                        } else if (currentHeight !== this.lastHeight && currentHeight > 0) {
                                            const newWidth = Math.round(currentHeight * this.aspectRatio);
                                            if (!isNaN(newWidth) && newWidth > 0) {
                                                widthWidget.value = newWidth;
                                                this.lastWidth = newWidth;
                                            }
                                            this.lastHeight = currentHeight;
                                        }
                                        
                                        this.lastCustomSize = {
                                            width: parseInt(widthWidget.value) || 1328,
                                            height: parseInt(heightWidget.value) || 1328
                                        };
                                        
                                        this.isUpdating = false;
                                    }
                                }
                            }
                        }, 100);
                        
                        applyAspectOptionsByMode(modeWidget.value, aspectRatioWidget);
                        
                        const isCustom = modeWidget.value === "Custom Size";
                        this.repositionCustomControls(isCustom);
                        
                        this.callback();
                    }
                }, 100);
                
                return r;
            };
        }
    },
});